quickwit/quickwit-common/src/rate_limited_tracing.rs

```diff
@@ -17,35 +17,75 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
+// TODO coasetime has a recent() instead of now() which is essentially free (atomic read instead of
+// vdso call), but needs us to spawn a future/thread updating that value regularly
+
 #[macro_export]
 macro_rules! rate_limited_tracing {
     ($log_fn:ident, limit_per_min=$limit:literal, $($args:tt)*) => {{
-        use ::std::sync::atomic::{AtomicU32, Ordering};
-        use ::std::sync::Mutex;
-        use ::std::time::{Instant, Duration};
+        use ::std::sync::atomic::{AtomicU64, Ordering};
+
+        use $crate::rate_limited_tracing::coarsetime::{Instant, Duration};
 
-        static COUNT: AtomicU32 = AtomicU32::new(0);
-        // we can't build an Instant from const context, so we pinitialize with a None
-        static LAST_RESET: Mutex<Option<Instant>> = Mutex::new(None);
+        //  This is treated as 2 u32: upper bits count "generation", lower bits count number of
+        //  calls since LAST_RESET. We assume there won't be 2**32 calls to this log in ~60s.
+        //  Generation is free to wrap arround.
+        static COUNT: AtomicU64 = AtomicU64::new(0);
+        const MASK: u64 = 0xffffffff;
```

Comment:
One way to make that code more readable and testable:

```rust
#[derive(Clone, Copy)]
struct Count {
    generation: u32,
    call_counts: u32, 
}

impl From<u64> for Count {
}


impl Into<u64> for Count {
}
```

if you inline the conversion method, it will be 0 cost to use only one of the two values.